h/nipsa/services.py

```diff
@@ -13,21 +13,23 @@ class NipsaService(object):
 
     def __init__(self, session):
         self.session = session
+        self._flagged_userids = None
 
     @property
-    def flagged_users(self):
+    def flagged_userids(self):
         """
-        A list of all the NIPSA'd users.
+        A list of all the NIPSA'd userids.
 
-        :rtype: list of unicode strings
+        :rtype: set of unicode strings
         """
-        return self.session.query(User).filter_by(nipsa=True)
+        if self._flagged_userids is None:
+            query = self.session.query(User).filter_by(nipsa=True)
```

Comment:
We can further optimize this by adding an `expression` property to the `UserIDComparator`, and then only querying for the `userid` "column":
```python
class UserIDComparator(Comparator):
    # ...
    @property
    def expression(self):
        return sa.sql.functions.concat(u'acct:', self.username, u'@', self.authority)

query = self.session.query(User.userid).filter_by(nipsa=True)
```

Which then runs the following query:
```
SELECT concat(%(concat_1)s, "user".username, %(concat_2)s, "user".authority) AS userid
FROM "user"
WHERE "user".nipsa = true
2016-12-21 16:54:15,269 [27855] [sqlalchemy.engine.base.Engine:INFO] {'concat_1': u'acct:', 'concat_2': u'@'}
```